slight refactor in `unexpand` ; fix and optimize `factor` ; fix `touch`, `test`, `tac`, `shuf`, `sleep` ; merge PR from @ctjhoa for `cksum`

kwantam opened the pull request:
This PR does the following:

1. a slight cleanup in `unexpand` that gets rid of redundant logic.
2. cherry-pick @wikol's PR #531 which adds Pollard's rho implementation to `factor`
3. Adds fast divisibility check and accompanying tables in `factor`, along with a generator for those tables.
4. Adds special cased arithmetic functions in `factor` to prevent overflow when input is larger than 2^63. This means we can always use Pollard's rho rather than falling back to trial division.
5. Adds three tests for `factor`.
6. Fixes `touch`.
  * Note that this utility does not work correctly when the `-h` option is supplied, as it uses Rust's `set_file_times()` stdlib function, which follows symlinks.
  * Fixing this requires using `lutimes`, but according to the man page that is only available on BSDs and Linux, so this will require a platform-specific solution.
7. Fixes `test`.
  * Parens remain unimplemented.
8. Fixes `tac`
9. Fixes `shuf`
10. Fixes `sleep`
11. Pulls in @ctjhoa's `cksum` PR from kwantam/coreutils#1

Closes #531 via cherry-pick.

Arcterus commented on src/factor/factor.rs:
```diff
@@ -12,66 +16,174 @@
+        x = rho_pollard_pseudorandom_function(x, a, b, num);
+        y = rho_pollard_pseudorandom_function(y, a, b, num);
+        y = rho_pollard_pseudorandom_function(y, a, b, num);
+        let d = gcd(num, max(x,y) - min(x,y));
```
I'd prefer that there be spaces after the commas here.

Arcterus commented on src/factor/factor.rs:
```diff
@@ -12,66 +16,174 @@
+    }
+    let divisor = rho_pollard_find_divisor(num);
+    rho_pollard_factor(divisor, factors);
+    rho_pollard_factor(num/divisor, factors);
```
Space between `num` and `/` and `/` and `divisor`.

Arcterus replied:
There are a few more spots like this that I haven't commented on.  Can you fix those as well?

Arcterus commented on src/factor/gen_table.rs:
```diff
@@ -0,0 +1,169 @@
+
+    let m = n;
+    Sieve::new()
+        .scan((0,3), move |st,x| {
```
Space between `,` and `x`.

Arcterus commented on src/factor/gen_table.rs:
```diff
@@ -0,0 +1,169 @@
+            let (count, mut cols) = *st;
+            if count < m {
+                // format the table
+                let outstr = format!("({},{},{}),", x, inv_mod_u64(x).unwrap(), MAX_U64/x);
```
Space between `MAX_U64` and `/`, and `/` and `x`.

Arcterus commented on src/factor/gen_table.rs:
```diff
@@ -0,0 +1,169 @@
+            } else {
+                None
+            }
+        }).take(m+1).count();
```
Space between `m` and `+`, and `+` and `1`.

Arcterus commented on src/factor/factor.rs:
```diff
@@ -12,66 +16,174 @@
-use std::old_io::stdio::stdin_raw;
+use numeric::*;
+use prime_table::{P_INVS_U64, NEXT_PRIME};
+use std::cmp::{max,min};
```
Space after comma.

Arcterus commented:
Everything looks great other than the minor spacing issues. :)

kwantam commented:
@Arcterus thanks for the review. I've taken care of the issues you pointed out, and I'm pretty sure I found the rest of them as well.

wikol commented on src/factor/factor.rs:
```diff
@@ -12,66 +16,174 @@
-        while num % i == 0 {
-            num /= i;
-            ret.push(i);
+    if is_prime(num) {
```
This might fail when num >= 2^63 as is_prime needs exponentiation modulo num.

kwantam replied:
> This might fail when num >= 2^63 as is_prime needs exponentiation modulo num.

I'll take a look into this. Thanks.

Heather commented:
@kwantam sadly this should be rebased again :(

kwantam commented:
@Heather no problem. Let me take a look at the potential issue @wikol pointed out above and then I'll rebase it and post another message to the effect that it's ready.

kwantam commented:
@Heather I have addressed @wikol's concern regarding overflow, which incidentally makes the worst case performance of `factor` substantially better because we can now always use Pollard's rho method. Also added a couple tests to exercise possible overflow paths. This is ready for a merge.

Heather commented:
Thank you, looks good to me.
